### ISSUE TITLE
Float params auto-converted to scientific notation, render pipeline spec invalid

### ISSUE THREAD

**Issue by jackwhelpton** [kubeflow/pipelines#3806]
### What steps did you take:

Defined a pipeline with float parameters:

```python
@dsl.pipeline(name="test", description="test")
def train_pipeline(
    schedule_1_learning_rate: float = 0.0002,
    schedule_2_learning_rate: float = 0.00002,
    schedule_3_learning_rate: float = 0.000002,
):
```

### What happened:

What gets compiled is:

```yaml
  arguments:
    parameters:
    - {name: schedule_1_learning_rate, value: '0.0002'}
    - {name: schedule_2_learning_rate, value: 2e-05}
    - {name: schedule_3_learning_rate, value: 2e-06}
```

This pipeline uploads fine, but when executing through the UI we receive:

```json
{"error":"Failed to create a new run.: Failed to fetch workflow spec.: Invalid input error: Please provide a valid pipeline spec","message":"Failed to create a new run.: Failed to fetch workflow spec.: Invalid input error: Please provide a valid pipeline spec","code":3,"details":[{"@type":"type.googleapis.com/api.Error","error_message":"Please provide a valid pipeline spec","error_details":"Failed to create a new run.: Failed to fetch workflow spec.: Invalid input error: Please provide a valid pipeline spec"}]}
```

If the largest value is used for all three parameters, no conversion occurs and the pipeline executes successfully.

### What did you expect to happen:

Pipeline executes successfully.

### Environment:

How did you deploy Kubeflow Pipelines (KFP)?

KFP version: 
Build commit: 9c16e12

KFP SDK version: 
kfp                      0.5.1
kfp-server-api           0.3.0

### Anything else you would like to add:
[Miscellaneous information that will assist in solving the issue.]

/kind bug

**Comment by Ark-kun** [comment]
Thank you for reporting. I'll fix this bug.

**Comment by TheFreaky** [comment]
Faced with same issue.

I found the reason for this behavior:

```python
>>> from kfp import dsl, compiler
... @dsl.pipeline(name="test", description="test")
... def train_pipeline(
...     schedule_1_learning_rate: float = 0.0002,
...     schedule_2_learning_rate: float = 0.00002,
...     schedule_3_learning_rate: float = 0.000002,
... ):
...     pass
... 
>>> import inspect
>>> inspect.getfullargspec(train_pipeline)
FullArgSpec(args=['schedule_1_learning_rate', 'schedule_2_learning_rate', 'schedule_3_learning_rate'], varargs=None, varkw=None, defaults=(0.0002, 2e-05, 2e-06), kwonlyargs=[], kwonlydefaults=None, annotations={'schedule_1_learning_rate': <class 'float'>, 'schedule_2_learning_rate': <class 'float'>, 'schedule_3_learning_rate': <class 'float'>})
>>> inspect.getfullargspec(train_pipeline).defaults
(0.0002, 2e-05, 2e-06)
>>> import yaml
>>> yaml.dump({'schedule_2_learning_rate': '2e-05'})
'schedule_2_learning_rate: 2e-05\n'
>>> yaml.dump({'schedule_2_learning_rate': '2.e-05'})
"schedule_2_learning_rate: '2.e-05'\n"
>>>
>>> # inspect.getfullargspec() retrieve float in form `2e-05`, but PyYAML requires `2.e-05`.
>>>
>>> import re
>>> yaml.Dumper.add_implicit_resolver(
...             u'tag:yaml.org,2002:float',
...             re.compile(u'''^(?:
...              [-+]?(?:[0-9][0-9_]*)\\.[0-9_]*(?:[eE][-+]?[0-9]+)?
...             |[-+]?(?:[0-9][0-9_]*)(?:[eE][-+]?[0-9]+)
...             |\\.[0-9_]+(?:[eE][-+][0-9]+)?
...             |[-+]?[0-9][0-9_]*(?::[0-5]?[0-9])+\\.[0-9_]*
...             |[-+]?\\.(?:inf|Inf|INF)
...             |\\.(?:nan|NaN|NAN))$''', re.X),
...             list(u'-+0123456789.'))
>>> yaml.dump({'schedule_2_learning_rate': '2e-05'})
"schedule_2_learning_rate: '2e-05'\n"
```

### Environment:

KFP version:
Build commit: 1168b08

KFP SDK version:
kfp 1.0.0rc5
kfp-server-api 1.0.0rc5

**Comment by aleksandra-sp** [comment]
In which release will be this be included? I am working on kfp 1.0.4 and the issue is still there:
`@dsl.pipeline(name="test", description="test")
def train_pipeline(
    adam_epsilon: float = 1e-8,
):`

What gets compiled:
`{name: adam_epsilon, value: 1e-08}`

the same happens when instead of passing 1e-8, I pass 0.00000001.

**Comment by Bobgy** [comment]
I think the fix was cherry picked to 1.0.x, so it should be there.
/assign @Ark-kun 
Can you confirm why this is still a problem?

**Comment by alvercau** [comment]
Hi

Any update on this? It's still not fixed, using version 1.2 now.

**Comment by ulrikpl** [comment]
This issue is still not fixed in KFP 1.4. So far I have gotten around it by using decimal notation in quotes. It doesn't seem to matter that component.yaml still describes the argument as 'Float'

(pipeline.py -> pipeline.yaml)
5e-04 -> 0.0005 (works)
5e-05 -> 5e-05 (fails)
'5e-04' -> 5e-04 (fails)
0.00005 -> 5e-05 (fails)
'0.00005' -> 0.00005 (works)
5e-05 -> 5e-05 -> '5e-05' (works) [last step is manuel edit of pipeline.yaml]

**Comment by MrMwalker2** [comment]
Had the same error when I changed the learning rate parameter of my pipeline from 0.0001 to 0.00001. It took me quite a while to realize that this was the cause of the problem.

So 0.0001 works fine, but 0.00001 doesn't.

Using Kubeflow 1.2 on AKS
